Dev guide contextualization | Content release #1 (#2327)
temporalio/documentation: docs-src/dev-guide/major-components.md

---
id: major-components
title: Temporal SDKs major components
description: Temporal SDKs provide developers with a Temporal Client to communicate with a Temporal Cluster, APIs to define Workflows, APIs to configure and start Workers, APIs to customize Activity Executions, and a common library for code that’s used across your application.
sidebar_label: Major components
tags:
  - temporal sdk
  - explanation
  - components
---

**What are the major components of Temporal SDKs?**

> Temporal SDKs offer developers with the following:

Reviewer: Temporal SDKs offer developers ~~with~~ the following:

- A Temporal Client to communicate with a Temporal Cluster

> - APIs to develop application code Workflows & Activities

Author: Workflows and Activities should probably have been in paranthesis here.
- APIs to develop application code (Workflows and Activities).

- APIs to configure and run Workers

> - A common library for code that’s used across the Client, Worker, and Workflow

Reviewer: This is slightly confusing because it reads like it's describing something distinct from the other bulletpoints, but I don't think it actually is.

![Temporal SDK components create a runtime across your environment and a Temporal Cluster](/diagrams/temporal-sdk-components.svg)

Let’s break down each one.

### Temporal Client

> A Temporal Client acts as a bridge of communication between the systems outside of Temporal and your application.

Reviewer: "outside of Temporal" seems wrong. How about

> A Temporal Client acts as the bridge for communication between your application and the Temporal server.

The Client performs key functions that facilitate the execution of, management of, and communication with Workflows.

The following code is an example using the Go SDK.
It showcases how to initialize a Temporal Client, create a connection to a local Temporal Cluster, and start a Workflow Execution:

:::caution Do not copy and use code

The following code is for example purposes only.
For tested code samples and best practices, use your preferred language SDK's developer's guide.

:::

```go
package main

import (
	"context"

	"go.temporal.io/sdk/client"
)

func main() {
  // Temporal Client setup code
	c, err := client.NewClient(client.Options{})
	if err != nil {
		log.Fatalln("Unable to create client", err)
	}
	defer c.Close()
  // Prepare Workflow option and parameters
	workflowOptions := client.StartWorkflowOptions{
		ID:        "loan-application-1",
		TaskQueue: "loan-application-task-queue",
	}
	applicantDetails := ApplicantDetails{
		// ...
	}

>   // Start the Workflow

Reviewer 2: This comment is improperly indented, as are the comments above it (lines 50 and 56)

	workflowRun, err := c.ExecuteWorkflow(context.Background(), workflowOptions, "loan-application-workflow", applicantDetails)
	if err != nil {
		// ...
	}
	// ...
}
```

Developers can then use the Client as the main entry point for interacting with the application through Temporal.

> From here, developers can now start or Signal Workflows, query the Workflow state etc.

Reviewer 3: ```suggestion
Using that Client, developers may for example start or signal Workflows, query a Workflow state, etc.
```

Author: 👍

> We can see in the example above how the developer has used the provided start method to start the Workflow upon the running of the function.

Reviewer 2: > how the developer has used the provided start method

It was unclear to me what you mean by "provided start method." Are you referring to the entire program (which I assume would be named `start/main.go`), the `client.ExecuteWorkflow` call, or something else? Regardless, I would avoid the word "method" since that may be interpreted in the object oriented way. Maybe "has used provided means of starting the Workflow" would be more clear.

Reviewer 3: > to start ~the~ Workflow ~upon the running of the function~.

to start **a** Workflow.

### APIs to Develop Workflows

> Workflows are defined as code as either a function or an object method, depending on the language.

Reviewer: How about

> Workflows are defined as code: either a function or an object method, depending on the language.

For example, the following is a valid Temporal Workflow in Go:

:::caution Do not copy and use code

The following code is for example purposes only.
For tested code samples and best practices, use your preferred language SDK's developer's guide.

:::

```go
func LoanApplication(ctx context.Context) (error) {
	// ...
	return nil
}
```

The Workflow code uses Temporal SDK APIs to orchestrate the steps of the application.

:::caution Do not copy and use code

The following code is for example purposes only.
For tested code samples and best practices, use your preferred language SDK's developer's guide.

:::

```go
func LoanApplication(ctx workflow.Context, input *LoanApplicationWorkflowInput) (*LoanApplicationWorkflowResult, error) {
	// ...
	var result activities.CreditCheckResult
	f := workflow.ExecuteActivity(ctx, a.CreditCheck, CreditCheckInput(*input))
	err := f.Get(ctx, &result)
	// ...
	// Return the results
	return &loanApplicationResults, nil
}
```

> A Workflow executes Activities (other functions), handles and sends messages (Queries, Signals, Updates), and interacts with other Workflows.

Reviewer 4: "other functions" is a bit vague to me..how about other functions that interact with external systems, functions that work with non-deterministic code etc.
